homeassistant/components/knx/__init__.py

```diff
@@ -58,6 +58,7 @@
 CONF_KNX_MCAST_PORT = "multicast_port"
 CONF_KNX_STATE_UPDATER = "state_updater"
 CONF_KNX_RATE_LIMIT = "rate_limit"
+CONF_KNX_LOG_DIRECTORY = "log_directory"
```

Comment:
I'm not sure about this one. The idea was to give the user the ability to log his/her KNX logs to a dedicated file (that will be created within the configured directory and will contain DEBUG logs (rotated every day and backup for 7 days)), giving us the possibility to analyse possible issues with the KNX integration easier since it's a complex protocol with many manufactures and devices.

However, when retesting this it appears that the log file only contains the same log levels than the real xknx logger which would not be benefical at all. Is there maybe a built in way to log integration logs on a lower log level (like DEBUG) to a file? I only saw that it's possible to change the log level.

Reply:
The basic design of python's logger is that the level set on the logger object happens first, then filters are applied, then the message gets passed to handlers on the current logger or any parent logger. 

The only way I know of to get the effect you want goes like this (which is much simpler than it sounds): You create a new logger hierarchy separate from the main one, so it can have separate log levels. You add a root level handler that forwards to the main hierarchy, at which point the main one's log levels are used. Indeed using this approach, I normally leave the levels unset in this second hierarchy, to ensure nothing gets filtered out before being forwarded to the main hierarchy. You can simply set the levels and any filtering you want on your handler instead.

To make things easy, one can do this as part of a custom module that exports a getLogger function, just like logging does. Basically drop the 32 line file shown in [this gist](https://gist.github.com/KevinCathcart/d26c7493896a729ddabf2e9a22ad647a) into your xknx project, and replace all `import logging` lines elsewhere in your project with `import xknxlogging`.  (Warning: I've modified this from my version so it is possible there is some minor bugs from typos, etc.)

Of course, perhaps some better way does exists, but I really don't see how it could, given how python's logging module works. Hope this helps.

Reply:
Thank you @KevinCathcart ! I'll give this a try and test it with the custom component